text/0000-dockerfiles.md

```diff
@@ -0,0 +1,113 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### App-specified Dockerfiles
+
+A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+### Platform-specified Dockerfiles
+
+The same Dockerfiles may be used to create new base images or modify existing base images outside of the app build process. For both app-specified and base-image-modifying Dockerfiles, any specified labels override existing values.
```

Comment:
What do you mean here `outside of the app build process` (that the `dockerfiles` part of a buildpack `builspack/<buildpackA/bin/` folder could also be used by a developer on their laptop to manually execute a `docker build` ?

Reply:
Two things I picked up from the working group discussion yesterday about this:

1. the Dockerfiles addition here is for app devs and for stack creators, it's not attempting to address buildpack authors. There is no mechanism for a buildpack author to provide a Dockerfile or to indicate that certain packages would need to be installed (I think #172 has some capabilities that a buildpack author could use to indicate requirements from the stack, that that #172 would install them but it allows you to indicate what is necessary from a stack to use your buildpacks).

2. "outside of the app build process" would refer to someone wanting to create a stack, so perhaps using `pack` cli to `pack stack create` or something like that (obviously other platforms could support similar capabilities).

Reply:
> 1. There is no mechanism for a buildpack author to provide a Dockerfile

Who will then provide such `run.Dockerfile` and `build.Dockerfile` files ?

Reply:
The two methods outlined in the RFC here are 1.) app devs and 2.) stack creators.

For apps devs, this RFC says `A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory.` so those would sit in a similar location as project.toml.

For stack creators, the location isn't tied to anything. You could manage them however you like. You'd then end up using a tool like `pack` to create the stack based on your customized docker files. Maybe `pack stack create`, I'm not sure if that was set in the RFC.

Reply:
@dmikusa-pivotal this thread is a bit old, but I think having hooks interact with the build plan now provides a way for buildpack authors to get things they need from hooks. A buildpack could indicate that it `requires` a dependency and trust that a hook will be there to `provide` it.